src/main/java/de/tum/in/www1/exerciseapp/service/CourseService.java

```diff
@@ -105,4 +103,68 @@ public void delete(Long id) {
         List<Course> courses = courseRepository.findAll();
         return courses.stream().map(c -> c.getTeachingAssistantGroupName()).collect(Collectors.toList());
     }
+
+    /**
+     * Getting a Collection of Results in which the average Score of all taken Exercises, is mapped to a userId for the instructor dashboard
+     *
+     * @param id the courseId
+     *
+     * @return The resultId refers in this case to the UserID to whom the overallScore belongsTo
+     */
+    @Transactional(readOnly = true)
+    public Collection<Result> getAllSummedOverallScoresOfCourse(Long id){
+        Course course = findOne(id);
+        Set<Exercise> exercisesOfCourse = course.getExercises();
+        HashMap<Long, Result> allOverallScoresOfCourse = new HashMap<>();
+
+        for(Exercise exercise : exercisesOfCourse){
+            Set<Participation> participations = exercise.getParticipations();
+            for (Participation participation : participations) {
+
+                long studentID = participation.getStudent().getId();
+                Result bestResult = bestResultScoreInParticipation(participation);
+                //setting student id in Result to refer to the student later in the webapp
+                bestResult.setId(studentID);
+
+                //if student already appeared once
+                if(allOverallScoresOfCourse.containsKey(studentID)){
+                    long oldScore = allOverallScoresOfCourse.get(studentID).getScore();
+                    bestResult.setScore(oldScore+bestResult.getScore());
+                    allOverallScoresOfCourse.remove(studentID);
+                }
+
+                allOverallScoresOfCourse.put(studentID, bestResult);
+            }
+        }
+        return allOverallScoresOfCourse.values();
+    }
+
+
+    /**
+     * Find the best Result in a Participation
+     *
+     * @param participation the participation you want the best result from
+     */
+    @Transactional(readOnly = true)
+    public Result bestResultScoreInParticipation(Participation participation) {
```

Comment:
I wanted to move this function either to participationService or ResultService. Or is it fine to leave it there

Reply:
I think it is ok to leave it here, but I would change the implementation: We always use the last result (according to the result.completionDate) within the exercise due date. This is an edge case, but it might happen that a student makes the solution worse and less test cases pass, which would lower the score. Then the latest results before the due date is the valid one. If there would be results after the due date, we would ignore them.

Reply:
Just to clarify, just look out to not check for any results after the due date?

Reply:
I did the check for the date hope I understood it correctly (see line 158)